src/compiler/checker.ts

```diff
@@ -5880,6 +5880,24 @@ namespace ts {
                     typeParameters = signature.typeParameters && signature.typeParameters.map(parameter => typeParameterToDeclaration(parameter, context));
                 }
 
+                // For regular function/method declarations, the enclosing declaration will already be signature.declaration,
+                // so this is a no-op, but for arrow functions and function expressions, the enclosing declaration will be
+                // the declaration that the arrow function / function expression is assigned to, which is not the right scope.
+                //
+                // If the parameters or return type include "typeof globalThis.paramName", using the wrong scope will lead
+                // us to believe that we can emit "typeof paramName" instead, even though that would refer to the parameter,
+                // not the global. Make sure we are in the right scope by changing the enclosingDeclaration to the function.
+                //
+                // We check that this signature's declaration is within the existing enclosing declaration so we don't pull
+                // in types from other files; if it is outside of the enclosing declaration, then we shouldn't be able to
+                // access those parameter types via typeof anyway. The declaration should also not be in a JS file; those can
+                // only be typed via JSDoc, but JSDoc annotations appear outside of the function and therefore are not in
+                // the function's scope, unlike true type annotations.
+                const saveEnclosingDeclaration = context.enclosingDeclaration;
+                if (signature.declaration && !isInJSFile(signature.declaration) && saveEnclosingDeclaration && findAncestor(signature.declaration, n => n === saveEnclosingDeclaration)) {

```

Comment:
I don't really like using the declaration from the signature over the context passed in - it may be in an entirely different file from the context where the signature is being emitted to, so will have potentially entirely incorrect results for things-that-aren't-immediately-parameters/type-parameters-of-the-signature. Instead we probably need to track the parameters we're serializing as we recur through multiple signatures - we already do this for type parameters in some cases. Similarly, we manufacture a fake node to use as a context in some places in the JS declaration builder that handles blocking/preserving scope variables like you'd want, since the real node doesn't exist (like how a `module.export.x = expr` needs a fake `module.exports` namespace declaration - passing in better initial contexts like this also sometimes an option.

Reply:
I was also concerned about this (it felt wrong entirely ignoring the context), but wasn't sure what to do instead. I'll try to come up something more like those other methods.

Reply:
> it may be in an entirely different file from the context where the signature is being emitted to, so will have potentially entirely incorrect results for things-that-aren't-immediately-parameters/type-parameters-of-the-signature (meaning it may allow references to locals in a separate file as though they're locals in the emitted file!)

Actually, just to clarify this, it was my impression that the ancestor check I've done here avoided that problem because it verifies that the current signature is a descendent of the current enclosing declaration.

This behavior is already not legal for declarations in other files because this involves `typeof` on the function itself. You can't directly do a `typeof` for a parameter that comes from an entirely different file (or even a different scope). The `findAncestor` check is what fixed that bug when I was working on this particular fix as there were some tests that did exhibit the bad behavior.

Reply:
If the function comes from within another function, but is being serialized in the top level scope of a js file, the scopes in between are still inaccessible. So some scope structure like
```
function f() {
  return g;
  /** @returns {typeof h} */
  function g() { return h; }
  function h() {}
}

module.exports = f;
```
Though probably with parameter references to inner functions in some way.

Reply:
I'm not sure if this is quite what you were suggesting, but the following doesn't work with my PR:

```ts
export function foo() {
    const bar = (isNan: typeof globalThis.isNaN): typeof globalThis.isNaN => isNan

    function baz() {
        return bar;
    }

    return baz;
}
```

[Playground Link](https://www.staging-typescript.org/play?ts=4.8.0-pr-49627-10&ssl=11&ssc=2&pln=3&pc=1#code/FAehAIHUCcHsDsDmAucBnAFrArgGwCboCGAnuAAYAuJADgKawBm4iusARkbgCoYCWaAHQCAckRHlwdAG51oJAO4Y5dcH3jhKy8PkGU0wOgA8asaJXCNs8AMaU+CS7FgAKAJTgA3sHC-wNhDQLTmhwAF5wF1EieFRqeiYWNk4efiFokTc42gZmVg4uXgFhNDERcIA+NVKY4B8-K1t7R04AL3cver8-aDpKbGgNEIBuLvAAXzru3v7B8DbRycMTMwsA+CDwIgBGcMiM7IS85MK0krKszRzE-JSi9JrysKqM0eNTc39AiyIAJj2oo9DrkkgVUsUMgAaeZEaAAfmBNxO4IeF0RxzB93O4kqMNCcLh1TKbxWn3WmyIAGYAQdwPBsABbdhyaEhdGgu5nDKXeIg26nCGPXEjZYfNbfLYAFhpQLpjOZ0B51wxnMFZVx-JR2JEwyAA)

I'm not sure I can actually produce a counterexample in JS, though; JSDoc is exempt from this PR just because the scope accessible in a JSDoc comment is outside the declaration, not inside of it.

Reply:
Hm, probably not what you meant; I think this just fails because the enclosing declaration is not a parent of the signature (which this PR used to work around things).